gadget-common/src/gadget/network/gossip.rs

```diff
@@ -592,14 +576,36 @@ impl<B: Block + 'static, KBE: KeystoreBackend> GossipHandler<B, KBE> {
             .insert(message.account_id, who);
     }
 
-    /// Called when peer sends us new signed DKG message.
-    async fn on_signed_dkg_message(&self, who: PeerId, message: GadgetProtocolMessage) {
+    /// Called when peer sends us new signed message.
+    async fn on_signed_message(&self, who: PeerId, mut message: GadgetProtocolMessage) {
         // Check behavior of the peer.
         self.logger.debug(format!(
-            "session {:?} | Received a signed DKG messages from {}",
+            "session {:?} | Received a signed messages from {}",
             message.associated_session_id, who,
         ));
 
+        let payload = message.payload;
+        let payload_and_sig: PayloadAndSignature = match Decode::decode(&mut payload.as_slice()) {
+            Ok(ret) => ret,
+            Err(e) => {
+                self.logger
+                    .error(format!("Failed to decode message payload: {e:?}"));
+                return;
+            }
+        };
+
+        if !self
+            .keystore
+            .pair()
+            .public()
+            .verify(&payload_and_sig.payload, &payload_and_sig.signature)
```

Comment:
Does this hash the payload under the hood? Is the signature over the hashed payload? We should document this and the functions a lot more please.

Reply:
We should do a simple ecdsa verify https://docs.rs/sp-io/latest/sp_io/crypto/fn.ecdsa_verify_prehashed.html and hash the payload using https://docs.rs/sp-io/latest/sp_io/hashing/fn.keccak_256.html. Same when we do signing.
Your implementation here will fail since it will try to verify that this message signed by us! Which is not the case, since this message is received from the network. We have the `who`'s peer id, which is mapped to their account id (public key).

Reply:
I will refactor this part @tbraun96 dw

Reply:
> Your implementation here will fail since it will try to verify that this message signed by us! Which is not the case, since this message is received from the network. We have the who's peer id, which is mapped to their account id (public key).

Correct, my mistake. In this case, get their public key and verify the signature. There's no need for hashing as both the `verify` and the `sign` use blake2b under the hood before actually signing/verifying the data. No need for double hashing with keccak.